src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -94,102 +94,173 @@ class MetricsBucket(TypedDict):
+    metrics_msg: Message[KafkaPayload]
+
+
 class BillingTxCountMetricConsumerStrategy(ProcessingStrategy[KafkaPayload]):
     """A metrics consumer that generates a billing outcome for each processed
     transaction, processing a bucket at a time. The transaction count is
     computed from the amount of values from `d:transactions/duration@millisecond`
     buckets.
     """
 
-    #: The ID of the metric used to count transactions
-    metric_id = TRANSACTION_METRICS_NAMES["d:transactions/duration@millisecond"]
+    counter_metric_name = "d:transactions/duration@millisecond"
 
     def __init__(
         self,
-        commit: Callable[[Mapping[Partition, Position]], None],
-        max_batch_size: int,
-        max_batch_time: int,
+        commit: Commit,
+        max_buffer_size: int,
     ) -> None:
-        self.__commit = commit
-        self.__max_batch_size = max_batch_size
-        self.__max_batch_time = timedelta(milliseconds=max_batch_time)
-        self.__messages_since_last_commit = 0
-        self.__last_commit = datetime.now()
-        self.__ready_to_commit: MutableMapping[Partition, Position] = {}
-        self.__closed = False
+        self._closed: bool = False
+        self._commit = commit
+
+        self._producers: Dict[str, KafkaProducer] = {}
+        # Especially on incidents generating big backlogs, we must expect a lot
+        # of removals from the beginning of a big FIFO. A dequeue provides O(1)
+        # time on removing items from the beginning; while a regular list takes
+        # O(n).
+        self._ongoing_billing_outcomes: Deque[MetricsBucketBilling] = deque()
+        self._max_buffer_size = max_buffer_size
 
     def poll(self) -> None:
-        if self._should_commit():
-            self._bulk_commit()
+        while self._ongoing_billing_outcomes:
+            self._process_metrics_billing_bucket(timeout=None)
+
+    def _process_metrics_billing_bucket(
+        self, timeout: Optional[float] = None, force: bool = False
+    ) -> None:
+        """
+        Takes the first billing outcome from the queue and if it's completed,
+        commits the metrics bucket.
+
+        When a future has thrown an exception, logs it and commits the bucket
+        anyway. wait_time is used to wait on the future to finish; if `None` is
+        provided and the future is not completed, no action is taken.
+        """
+        if len(self._ongoing_billing_outcomes) < 1:
+            return
+
+        billing_future, metrics_msg = self._ongoing_billing_outcomes[0]
+        if billing_future:
+            try:
+                if not timeout and not billing_future.done():
+                    return
+                billing_future.result(timeout)
+            except Exception:
+                logger.error(
+                    "Async future failed in billing metrics consumer.",
+                    exc_info=billing_future.exception(),
+                    extra={"offset": metrics_msg.offset},
+                )
```

Comment:
Fyi, this will provide pretty much the same delivery guarantee you had with `KafkaPublisher` as this code ignores failures in the producer (albeit it logs them).
Is this what you intended ? 

The PR description mentions:

> The consumer keeps track of the futures, and only when a future is resolved the associated metric bucket is committed. This guarantees billing outcomes are produced exactly once during execution and regular shutdown. On crashes (e.g. an exception is raised while committing), some billing outcomes may be duplicated (i.e. at least once delivery).

This implementation gives you at most once (with the exceptions of hard crashes) as it ignores the failures.